connection_test.go

```diff
@@ -186,3 +190,105 @@ func TestCloseIgnoreSessionGone(t *testing.T) {
 		t.Error("Close should let go session gone error")
 	}
 }
+
+// test
+func TestFetchResultByQueryID(t *testing.T) {
+	fetchResultByQueryID(t, nil, nil)
+}
+
+func TestFetchRunningQueryByID(t *testing.T) {
+	fetchResultByQueryID(t, returnQueryIsRunningStatus, nil)
+}
+
+func TestFetchErrorQueryByID(t *testing.T) {
+	err := &SnowflakeError{
+		Number: ErrQueryReportedError}
+	fetchResultByQueryID(t, returnQueryIsErrStatus, err)
+}
+
+func customGetQuery(ctx context.Context, rest *snowflakeRestful, url *url.URL,
+	vals map[string]string, _ time.Duration, jsonStr string) (*http.Response, error) {
+	if strings.Contains(url.Path, "/monitoring/queries/") {
+		return &http.Response{
+			StatusCode: http.StatusOK,
+			Body:       ioutil.NopCloser(strings.NewReader(jsonStr)),
+		}, nil
+	}
+	return getRestful(ctx, rest, url, vals, rest.RequestTimeout)
+}
+
+func returnQueryIsRunningStatus(ctx context.Context, rest *snowflakeRestful, fullURL *url.URL,
+	vals map[string]string, duration time.Duration) (*http.Response, error) {
+	var jsonStr = `{"data" : { "queries" : [{"status" : "RUNNING", "state" : "FILE_SET_INITIALIZATION",
+        "errorCode" : 0, "errorMessage" : null}] }, "code" : null, "message" : null, "success" : true }`
+	return customGetQuery(ctx, rest, fullURL, vals, duration, jsonStr)
+}
+
+func returnQueryIsErrStatus(ctx context.Context, rest *snowflakeRestful, fullURL *url.URL,
+	vals map[string]string, duration time.Duration) (*http.Response, error) {
+	var jsonStr = `{"data" : { "queries" : [{"status" : "FAILED_WITH_ERROR", "errorCode" : 0, "errorMessage" : ""}] },
+       "code" : null, "message" : null, "success" : true }`
+	return customGetQuery(ctx, rest, fullURL, vals, duration, jsonStr)
+}
+
+// this function is going to: 1, create a table, 2, query on this table,
+//      3, fetch result of query in step 2, mock running status and error status of that query.
+func fetchResultByQueryID(t *testing.T, customget FuncGetType, expectedFetchErr *SnowflakeError) error {
+
+	config, _ := ParseDSN(dsn)
+	ctx := context.Background()
+	sc, err := buildSnowFlakeConn(ctx, *config)
```

Comment:
I noticed that this test function also does:
- buildSnowflakeConn
- authenticateWithConfig
- startHeartBeat

These seem to be doing the same as `driver.OpenWithConfig()`, so is this equivalent to:
```
conn, err := driver.OpenWithConfig(...)
sc := conn.(snowflakeConn)
if customget != nil {
  sc.rest.FuncGet = customget
}
```

If these two are equivalent, then I would prefer we just call `OpenWithConfig()` because that's what the normal client code path uses

Reply:
For connection_test or other tests, heart beat is not needed so I prefer keeping it current way and remove the heart beat part.